File: typescript/stage4.ts
import type { ListElement } from "./Element";


/**
 * sort a list of elements in ascending order with O(n) maximum time complexity.
 * @param elements list of elements to sort (sorting will modify the list).
 */
function sort(elements: ListElement[]): void {
  // yeah. you can do it. there is a hint in the documentation of the Element class.
  let length = 2 ** 16;
  let sortedElements: ListElement[][] = [];
  for (let i = 0; i < length; i++) {
    sortedElements[i] = [];
  }
  for (let element of elements) {
    let value = element.toInteger();
    sortedElements[value].push(element);
  }
  let i = 0;
  for (let equalElements of sortedElements) {
    elements.splice(i, equalElements.length, ...equalElements);
    i += equalElements.length;
  }
}


export { sort };

File: python/stage4.py
from Element import Element


def sort(elements: list[Element]):
    """
    sort a list of elements in ascending order with O(n) maximum time complexity.
    @param elements: list of elements to sort (sorting will modify the list).
    """
    # yeah. you can do it. there is a hint in the documentation of the Element class.
    sorted_elements = [[] for _ in range(2**16)]
    for element in elements:
        value = int(element)
        sorted_elements[value].append(element)
    i = 0
    for equal_elements in sorted_elements:
        elements[i:i + len(equal_elements)] = equal_elements
        i += len(equal_elements)
